### ISSUE TITLE
setInput() triggers ngOnChanges  and mark component when it should not

### ISSUE THREAD

**Issue by montella1507** [angular/angular#47472]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

I thought ComponentRef.setInput() will:

- set input
- mark component when neccessary
- make sure ngOnChanges is called when neccessary

However it is called when OnPush component is used and input doesn't change so i think it is not correct behavior.

![image](https://user-images.githubusercontent.com/10418323/190913411-17d5617d-d16f-45df-a6dc-7497518e0272.png)

https://user-images.githubusercontent.com/10418323/190913482-70c5da4c-1855-479f-a744-20fae0afff8e.mp4

### Please provide a link to a minimal reproduction of the bug
https://stackblitz.com/edit/angular-ivy-yo1znz?file=src/app/hello.component.ts

### Anything else?

Of course we can check it manually, but only "component.instance" is available for the developers from "PUBLIC API" - and that's problem in case of setters without getters etc.

**Comment by pkozlowski-opensource** [comment]
Currently the `setInput` is _not_ comparing prev / current values and this was conscious design decision.  Something we might revisit as I've heard this need expressed elsewhere.

**Comment by montella1507** [comment]
Hi  @pkozlowski-opensource , thanks for your answer. 

If i am correct, prev / current values are compared "SOMEWHERE" higher than on component level, right? In some kind of "component host container", there are stored previous inputs and ngOnChanges may be called  (and component is marked) based on the result of that compare. 

AFAIK we don't have "public" access to that place so we cannot do that (compare prev/current) with public api?

**Comment by pkozlowski-opensource** [comment]
`ngOnChanges` doesn't do prev / current comparison, this happens _before_ input values are updated (and hence it is not implemented in the current version of the API). So it is not a question of the public API availability. Currently you would have to remember prev values and do the comparison on your side. As I've mentioned, this is sth we might consider doing inside `setInput`, although it is not clear what the design should be and if this is the right direction at all...

**Comment by montella1507** [comment]
So i have to do something like 

```
_previousName = undefined;

setName(newName: string) {
  if (newName !== this._previousName) {
   this._previousName = newName;
    this.setInput('name', newName);
  }
}

```
for every input (in the universal way  - i can use reflectComponentType() api to get list of the inputs though)

**Comment by pkozlowski-opensource** [comment]
So this is a feature request I've heard before and we might need to do it for router anyway so marking it as a feature request (to be clear: it is about actually invoking `setInput` _only_ if the value being set is different from the previously set value).

I'm not sure yet how this should be implemented and if this should end up in core or not. But it is a legitimate feature request that should be implementable within the current design.

**Comment by montella1507** [comment]
@pkozlowski-opensource  @jessicajaniuk  Can we at least mention this not so correct behavior in documentation? 

There is 

`Using this method will properly mark for check component using the OnPush change detection strategy. It will also assure that the OnChanges lifecycle hook runs when a dynamically created component is change-detected.`

And this is not correct as far as these 2 are used when they shouldn't be used.

**Comment by pkozlowski-opensource** [comment]
This issue was fixed by #49607
